sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -14,197 +16,180 @@ from ...cluster._hierarchical_fast cimport UnionFind
+            if mutual_reachability_distance < next_node_min_reach:
+                min_reachability[j] = mutual_reachability_distance
+                current_sources[j] = current_node
+                if mutual_reachability_distance < new_reachability:
+                    new_reachability = mutual_reachability_distance
+                    source_node = current_node
                     new_node = j
             else:
-                if right_value < new_distance:
-                    new_distance = right_value
-                    source_node = right_source
+                if next_node_min_reach < new_reachability:
+                    new_reachability = next_node_min_reach
+                    source_node = next_node_source
                     new_node = j
 
-        result[i - 1, 0] = <double> source_node
-        result[i - 1, 1] = <double> new_node
-        result[i - 1, 2] = new_distance
+        mst[i - 1].current_node = source_node
+        mst[i - 1].next_node = new_node
+        mst[i - 1].distance = new_reachability
         current_node = new_node
 
-    return result_arr
+    return mst
 
 @cython.wraparound(True)
-cpdef cnp.ndarray[cnp.double_t, ndim=2] label(cnp.double_t[:,:] L):
+cpdef cnp.ndarray[cnp.float64_t, ndim=2, mode='c'] label(MST_edge_t[::1] mst):
 
     cdef:
-        cnp.ndarray[cnp.double_t, ndim=2] result_arr
-        cnp.double_t[:, ::1] result
+        cnp.ndarray[cnp.float64_t, ndim=2, mode='c'] single_linkage
 
-        cnp.intp_t N, a, aa, b, bb, index
-        cnp.double_t delta
+        # Note mst.shape[0] is one fewer than the number of samples
+        cnp.intp_t n_samples = mst.shape[0] + 1
+        cnp.intp_t current_node_ancestor, next_node_ancestor
+        cnp.intp_t current_node, next_node, index
+        cnp.float64_t distance
 
-    result_arr = np.zeros((L.shape[0], L.shape[1] + 1))
-    result = (<cnp.double_t[:L.shape[0], :4:1]> (
-        <cnp.double_t *> result_arr.data))
-    N = L.shape[0] + 1
-    U = UnionFind(N)
+    single_linkage = np.zeros((n_samples - 1, 4))
+    U = UnionFind(n_samples)
 
-    for index in range(L.shape[0]):
+    for i in range(n_samples - 1):
 
-        a = <cnp.intp_t> L[index, 0]
-        b = <cnp.intp_t> L[index, 1]
-        delta = L[index, 2]
+        current_node = mst[i].current_node
+        next_node = mst[i].next_node
+        distance = mst[i].distance
 
-        aa, bb = U.fast_find(a), U.fast_find(b)
+        current_node_ancestor, next_node_ancestor = (
+            U.fast_find(current_node),
+            U.fast_find(next_node)
+        )
 
-        result[index][0] = aa
-        result[index][1] = bb
-        result[index][2] = delta
-        result[index][3] = U.size[aa] + U.size[bb]
+        single_linkage[i][0] = current_node_ancestor
+        single_linkage[i][1] = next_node_ancestor
+        single_linkage[i][2] = distance
+        single_linkage[i][3] = U.size[current_node_ancestor] + U.size[next_node_ancestor]
```

Comment:
This will eventually be converted to a structured array, however I'm leaving that for the cleanup/refactor of `_hdbscan/_tree.pyx` since that cython code will need to be made compatible at the same time.